src/imputors/drop.jl

```diff
@@ -1,73 +1,143 @@
-* We use `view`, but this will change the type of the `data` by returning a `SubArray`
-* We might be able to do something clever by:
-    1. reshaping the data to a vector
-    2. running `deleteat!` for the appropriate indices and
-    3. reshaping the data back to the desired shape.
+function impute!(table, imp::DropObs)
+    imp.context() do c
+        @assert istable(table)
+        rows = Tables.rows(table)
 
-# Arguments
-* `imp::Drop`: this `Imputor` method
-* `ctx::Context`: contextual information for missing data
-* `data::AbstractMatrix`: the data to impute
+        # Unfortunately, we'll need to construct a new table
+        # since Tables.rows is just an iterator
+        table = Iterators.filter(rows) do r
+            !any(x -> ismissing(c, x), propertyvalues(r))
+        end |> materializer(table)
 
-# Returns
-* `AbstractMatrix`: a new matrix with missing rows removed
-"""
-function impute!(imp::Drop, ctx::Context, data::AbstractMatrix)
-    ctx.num = size(data, 1)
-    mask = map(i -> !ismissing(ctx, data[i, :]), 1:size(data, 1))
-    return data[mask, :]
+        return table
+    end
+end
+
+
+struct DropVars <: Imputor
+    vardim::Int
+    context::AbstractContext
 end
 
 """
-    impute!(imp::Drop, ctx::Context, data::DataFrame)
+    DropVars(; vardim=2, context=Context())
+
 
-Finds the missing rows in the `DataFrame` and deletes them.
+Finds variables with too many missing values in a `AbstractMatrix` or `Tables.table` and
+removes them from the input data.
 
-NOTE: this isn't quite as fast as `dropnull` in `DataFrames`s as we're using an arbitrary
-`missing` function rather than using the precomputed `dt.isnull` vector of bools.
+# Keyword Arguments
+* `vardim=2::Int`: Specify the dimension for variables in matrix input data
+* `context::AbstractContext`: A context which keeps track of missing data
+  summary information
 
-# Arguments
-* `imp::Drop`: this `Imputor` method
-* `ctx::Context`: contextual information for missing data
-* `data::DataFrame`: the data to impute
+# Examples
+```jldoctest
+julia> using Impute: DropVars, Context, impute
 
-# Returns
-* `DataFrame`: our data with the missing rows removed.
+julia> M = [1.0 2.0 missing missing 5.0; 1.1 2.2 3.3 missing 5.5]
+2×5 Array{Union{Missing, Float64},2}:
+ 1.0  2.0   missing  missing  5.0
+ 1.1  2.2  3.3       missing  5.5
+
+julia> impute(M, DropVars(; vardim=1, context=Context(; limit=0.2)))
+1×5 Array{Union{Missing, Float64},2}:
+ 1.1  2.2  3.3  missing  5.5
+```
 """
-function impute!(imp::Drop, ctx::Context, data::DataFrame)
-    ctx.num = size(data, 1)
-    m = typeof(data).name.module
-    m.deleterows!(data, findall(r -> ismissing(ctx, r), m.eachrow(data)))
-    return data
+DropVars(; vardim=2, context=Context()) = DropVars(vardim, context)
+
+function impute!(data::AbstractMatrix, imp::DropVars)
+    return filtervars(imp, data) do var
+        try
+            imp.context() do c
+                for x in var
+                    ismissing(c, x)
```

Comment:
oh this is really uncofortable,
that `ismissing` mutates state.

It looks like this loops is doing nothing at first glance

Reply:
i am also confused by this